Update "Bootstrapping" chapter (#1225)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Bootstrapping/Index.rst

=============

TYPO3 CMS has a clean bootstrapping process driven mostly
by class :php:`\TYPO3\CMS\Core\Core\Bootstrap`. This class is initialized by
calling  :php:`Bootstrap::init()` and serves as an entry point for later calling
an application class, depending on several context-dependant constraints.

> Each application class registers Request Handlers to the TYPO3 Bootstrap to

Reviewer: ```suggestion
Each application class registers request handlers to
```

run a certain request type (e.g. eID or TSFE-logic, or AJAX requests in the Backend). Each application is handed
over the class loader provided by Composer.

Applications
============

There are four types of applications provided by the TYPO3 Core:

\\TYPO3\\CMS\\Frontend\\Http\\Application
-----------------------------------------

This class handles all incoming web requests coming through :file:`index.php`

> in the main directory. It handles all TSFE and eID requests.

Reviewer: ```suggestion
in the public web directory. It handles all regular page (TSFE) and eID requests.
```

It checks if all configuration is set, otherwise redirects to the TYPO3 Install
Tool.

\\TYPO3\\CMS\\Backend\\Http\\Application
----------------------------------------

This class handles all incoming web requests for any regular backend call
inside :file:`typo3/\*`.

> Its :php:`TYPO3\CMS\Backend\Http\RequestHandler` is used for all Backend

Reviewer: ```suggestion
Its :php:`TYPO3\CMS\Backend\Http\RequestHandler` is used for all backend
```

requests, including AJAX routes. If a get/post parameter "route" is set, the

> Backend Routing is called by the :php:`RequestHandler` and

Reviewer: ```suggestion
backend routing is called by the :php:`RequestHandler` and
```

> searches for a matching route inside the Router. The corresponding controller

Reviewer: ```suggestion
searches for a matching route inside the router. The corresponding controller
```

/ action is called then which returns the response.

The :php:`Application` checks if all configuration is set, otherwise it
redirects to the TYPO3 Install Tool.

\\TYPO3\\CMS\\Core\\Console\\CommandApplication
-----------------------------------------------

This class is the entry point for the TYPO3 command line for console commands.

> In addition to a simple Symfony command, this also sets up a CLI user.

Reviewer: ```suggestion
In addition to registering all available commands, this also sets up a CLI user.
```

> In addition to a simple Symfony command, this also sets up a CLI user.

Reviewer: What is the "simple Symfony command"?

Author: No idea, do you know? Or should we just remove the sentence? (Please note I only try to get this old PR merged, I didn't write the texts in the first place)

\\TYPO3\\CMS\\Install\\Http\\Application
----------------------------------------

The install tool :php:`Application` only runs with a very limited bootstrap
set up. The failsafe package manager does not take

> the :php:`ext_localconf.php` scripts of installed extensions into account.

Reviewer: ```suggestion
the :php:`ext_localconf.php` of installed extensions into account.
```

.. warning::

   This bootstrapping API is internal and may change at any time in the near future
   even in minor updates. It is thus discouraged to use it in third party code.
   Use this class only if other extensibility possibilities such as

>    :ref:`Hooks <hooks>`, Signals or :ref:`XCLASS <xclasses>`

Reviewer: ```suggestion
   :ref:`Events <eventdispatcher>`, :ref:`Hooks <hooks>`, or :ref:`XCLASS <xclasses>`
```

   are not enough to reach your goals.

Example of bootstrapping the TYPO3 Backend:

.. code-block:: php

   // Set up the application for the backend
   call_user_func(function () {

>        $classLoader = require dirname(__DIR__).'/vendor/autoload.php';

Reviewer: ```suggestion
       $classLoader = require dirname(__DIR__) . '/vendor/autoload.php';
```

=============

TYPO3 CMS has a clean bootstrapping process driven mostly
by class :php:`\TYPO3\CMS\Core\Core\Bootstrap`. This class is initialized by
calling  :php:`Bootstrap::init()` and serves as an entry point for later calling

> an Application class, depending on the TYPO3_MODE and several context-dependant

Reviewer: I think we should keep regular terms like `application` in lowercase unless we refer to a specific "thing" which will then very likely be a class name.

Also we should not refer to `TYPO3_MODE` anymore which is deprecated in TYPO3v11.

bootstrapping process managed by a dedicated API. This process is also
used in the frontend, but only the backend process is described here.

> Classes involved in the backend bootstrapping process are :php:`\TYPO3\CMS\Core\Core\Bootstrap`,

Reviewer: There are *many* more classes involved so what was the intention of mentioning these?

Author: Removed the sentence

The following steps are performed during bootstrapping.

1. Initialize the class loader
------------------------------

This defines which autoloader to use.

2. Run SystemEnvironmentBuilder
-------------------------------

The :php:`\TYPO3\CMS\Core\Core\SystemEnvironmentBuilder` is responsible for setting
up a system environment that is shared by all contexts (FE, BE, Install Tool and CLI).
This class defines a large number of constants and global variables. If you want
to have an overview of these base values, it is worth taking a look into the following methods:

-  :php:`SystemEnvironmentBuilder::defineTypo3RequestTypes()` defines the different
   constants for determining if the current request is a frontend, backend, cli,
   ajax or Install Tool request.

-  :php:`SystemEnvironmentBuilder::defineBaseConstants()` defines
   constants containing values such as the current version number,
   blank character codes and error codes related to services.

-  :php:`SystemEnvironmentBuilder::initializeEnvironment()` initializes the

>    Environment class that points to various parts of the TYPO3 installation like

Reviewer: ```suggestion
   :php:`Environment` class that points to various parts of the TYPO3 installation like
```

   the absolute path to the :file:`typo3` directory or the absolute
   path to the installation root.

-  :php:`SystemEnvironmentBuilder::calculateScriptPath()` calculates the script
   path. This is the absolute path to the entry script. This can be something
   like '.../public/index.php' or '.../public/typo3/index.php' for web calls,

>    or '.../bin/typo3' or similar for cli calls.

Reviewer: ```suggestion
   or '.../bin/typo3' or similar for CLI calls.
```

   defines special variables which contain, for example, the current time or
   a simulated time as may be set using the Admin Panel.

3.  Initialize bootstrap
------------------------

:php:`\TYPO3\CMS\Core\Core\Bootstrap` boots up TYPO3 and returns a container
that is later used to run an application. As a basic overview it does the
following:

-  :php:`Bootstrap::initializeClassLoader()` processes all the information
   available to be able to determine where to load classes from, including class
   alias maps which are used to map legacy class names to new class names.

-  :php:`Bootstrap::checkIfEssentialConfigurationExists()` checks if crucial
   configuration elements have been set. If that is not the case, the
   installation is deemed incomplete and the user is redirected to the Install Tool.

-  :php:`Bootstrap::createConfigurationManager()` creates the Configuration
   Manager which is then populated with the the main configuration ("TYPO3_CONF_VARS").

> -  :php:`$builder->createDependencyInjectionContainer()` creates a dependency
>    injection container which is later returned by :php:`Bootstrap::init()`.

Reviewer: ```suggestion
-  :php:`$builder->createDependencyInjectionContainer()` creates a :ref:`dependency
   injection <dependency-injection>` container which is later returned by :php:`Bootstrap::init()`.
```

> -  The Caching Framework and the Package Management are set up.

Reviewer: ```suggestion
-  The caching framework and the package management are set up.
```
